src/js/oncoprintruleset.js

```diff
@@ -935,11 +935,80 @@ var Rule = (function () {
     return Rule;
 })();
 
+// The GradientCategoricalRuleSet class delegates RuleSet API
+// calls to GradientRuleSet and/or CategoricalRuleSet classes.
+class GradientCategoricalRuleSet extends RuleSet {
+
+	constructor(params) {
+		super(params);
+		this.categoricalRuleSet = new CategoricalRuleSet(params);
+		this.gradientRuleSet = new GradientRuleSet(params);
+	}
+
+	// RuleSet API
+	apply(data, cell_width, cell_height, out_active_rules, data_id_key, important_ids) {
+
+		var dataRules = [];
+
+		// check the type of datum (categorical or gradient)
+		// and delegate the fetching of shape to the appropriate
+		// RuleSet type
+		for (var i = 0; i < data.length; i++) {
+			var datum = data[i];
+			if ( this.isCategorical(datum) ) {
+				dataRules.push( this.categoricalRuleSet.apply([datum], cell_width, cell_height, out_active_rules, data_id_key, important_ids)[0] );
+			} else if ( this.isGradient(datum) ) {
+				dataRules.push( this.gradientRuleSet.apply([datum], cell_width, cell_height, out_active_rules, data_id_key, important_ids)[0] );
+			} else {
+				dataRules.push([]);
+			}
+		}
+
+		return dataRules;
+	}
+
+	// RuleSet API
+	getRulesWithId(datum) {
+		var categoricalRules = this.categoricalRuleSet.getRulesWithId(datum);
+		var gradientRules = this.gradientRuleSet.getRulesWithId(datum);
+		var rules = categoricalRules.concat(gradientRules);
+		return this.removeNoDataDuplicates(rules);
+	}
+
+	// helper function
+	removeNoDataDuplicates(rules) {
```

Comment:
in my opinion a nicer way to do this would actually be to pass an argument to the rule set constructors to tell them not to create an NA rule, then add one here. that keeps with the convention in this file of having each rule set create an NA rule. what do you think?

Reply:
Phew, this took some further analysis from my part. I have implemented your suggestion (Gradient and Categorical RuleSet take a parameter that suppresses addition of the NA-rule by these classes), but devised a different strategy for NA-rule handling. The rationale is stated in the GradientCategoricalRuleSet class as follows:

> // For the GradientCategoricalRuleSet a datum must always have a 
> // value (for heatmap clustering) and can have a category attribute. 
> // A datum is 'NA' when not meeting the requirements for the GradientRuleSet.
> // To achieve correct evaluation, the CategoricalRuleSet is asked not to
> // contribute an `NA` rule.

So, for evaluation of data for a hybrid heatmap track the CategoricalRuleSet is asked to suppress its NA-rule, whereas the GradientRuleSet is asked to provide one.

> this.gradientRuleSet = new GradientRuleSet(params);
> this.categoricalRuleSet = new CategoricalRuleSet(params, omitNArule);

I think this solution is preferred over handling of the NA-rule by the new GradientCategoricalRuleSet class itself.

What is your opinion?

Reply:
I think its a very good solution